### PR TITLE
[eslint-plugin-react-hooks] Report constant constructions

### PR DESCRIPTION
**Pull request by captbaritone**
## Summary

This PR expands upon the existing logic which will report functions defined inside the render function as invalid hook dependencies.

We can now detect additional types of expressions which will result in values that will be referentially unique on each render, such as `[]` or `{}`.

```
function MyComponent() {
  const foo = ['a', 'b', c']; // <== This array is reconstructed each render
  const normalizedFoo = useMemo(() => foo.map(expensiveMapper), [foo]);
  return <OtherComponent foo={normalizedFoo} />
}
```

We also detect cases where an expression may conditionally result in a referentially unique on each render, such as `someArr ?? []` or `someObj || {}`.

```
function MyComponent({ nullableValue }) {
  const foo = nullableValue ?? ['a', 'b', c']; // <== This array is _potentially_ reconstructed each render
  const normalizedFoo = useMemo(() => foo.map(expensiveMapper), [foo]);
  return <OtherComponent foo={normalizedFoo} />
}
```

## Test Plan

`yarn jest`

### CODE REVIEW

File: packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js [hunk 1]
```diff
@@ -1381,50 +1392,97 @@ function collectRecommendations({
...
+      ) {
+        return 'conditional';
+      }
+      return null;
+    case 'LogicalExpression':
+      if (
+        getConstructionExpresionType(node.left) != null ||
+        getConstructionExpresionType(node.right) != null
+      ) {
+        return 'logical expression';
```
Comment (captbaritone) [inline packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js:1421]: For even better developer experience, we could try to blame the actual sub-expression that is a construction.

For example:

```
const foo = bar ?? [];
                   ^^ This is actually to blame
```

However, it would add some additional complexity.
```

**Comment by captbaritone** [comment]
FYI @gaearon who suggested we add this here rather than as a separate rule.

cc @dylanOshima who paired with me on the original, standalone, version of this rule.

**Comment by captbaritone** [comment]
The `ExhaustiveDeps` test was segfaulting on my work machine unless I omitted most of the tests, so I was only running my new tests (using `only: true`). They seem to run fine on my personal computer, so I'll fix up failing tests now.

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js [hunk 2]
```diff
@@ -5747,56 +5813,6 @@ const tests = {
...
-      errors: [
-        {
-          message:
-            "The 'handleNext' function makes the dependencies of useEffect Hook " +
-            '(at line 13) change on every render. To fix this, wrap the ' +
-            "'handleNext' definition into its own useCallback() Hook.",
-          // Normally we'd suggest moving handleNext inside an
-          // effect. But it's used more than once.
-          // TODO: our autofix here isn't quite sufficient because
-          // it only wraps the first definition. But seems ok.
```
Comment (captbaritone) [inline packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:5775]: I had decided at reassigned let/var values were too complex to reason about, so this case would no-longer be an error. How do we want to think about let/var declarations with reassignment?

I suppose we could say that if the initial assignment is a construction, then it's invalid, and if you later assign it another construction we won't try to catch that.

We could then safely suggest that (as in this case) you at least wrap your initial declaration in a useMemo/useCallback.

**Comment by captbaritone** [comment]
I got this working on VSCode with some actual production code and looked at a few examples. Some thoughts:

## Suggestions are helpful but clunky

If I have a conditional array construction in my render function the flow looks like this:

1. Start state:
  * `const foo = bar ?? []`;
2. See the error and apply the **automatic** suggestion to wrap in `useMemo`:
  * `const foo = useMemo(() => {return bar ?? [];});`
3. Realize that `useMemo` is not imported. **Manually** import that (auto import screwed up due to required/import migration)
  * `import {useMemo} from 'React';`
4. Lint complains about missing dependency array, but does not suggest a fix. **Manually** add an empty one:
  * `const foo = useMemo(() => {return bar ?? [];}, []);`
5. Lint informs me I'm missing a dependency `bar` and offers an **automatic** fix, which I take:
  * `const foo = useMemo(() => {return bar ?? [];}, [bar]);`

Overall this is clunky, but at least I get an error at each point and the error tells me what to do.

_Should out suggested fix include an empty dependency array?_ That changes the semantics, but gets the user closer to a fix, assuming they listen to other lint advice.

## Wrapping only the declaration may cause bugs if the value get mutated

Consider this case:

```
function Greeting({name}) {
  const myArr = ['Hello'];
  if (name) {
    myArr.push(name);
  }
  const greeting = useMemo(() => myArr.join(' '), [myArr]);
  return greeting;
}
```

Currently we offer an automatic suggestion to wrap the declaration in `useMemo`, but that would cause a bug:

```
function Greeting({name}) {
  const myArr = useMemo(() => ['Hello'];
  if (name) { // <== ERROR: `name` could change and `greeting` would not update.
    myArr.push(name);
  }
  const greeting = useMemo(() => myArr.join(' '), [myArr]);
  return greeting;
}
```

_Maybe we need to check for the case where the value gets mutated and not offer an automated fix in that case_

**Comment by gaearon** [comment]
>Should our suggested fix include an empty dependency array?

Can we suggested a filled-in array in the first place?

**Comment by gaearon** [comment]
>Currently we offer an automatic suggestion to wrap the declaration in useMemo, but that would cause a bug:

Yeah — this is really, really bad.

How do you feel about starting without an autofix at all? We can maybe add it later if it's too burdensome.

**Comment by captbaritone** [comment]
> How do you feel about starting without an autofix at all? We can maybe add it later if it's too burdensome.

Sounds good. Should we leave the autofix for functions on the assumption that it's not very common for folks to mutate functions?

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js (hunk 2 above)
Comment (gaearon) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:5775]: >I suppose we could say that if the initial assignment is a construction, then it's invalid, and if you later assign it another construction we won't try to catch that.

Seems reasonable.

**Comment by captbaritone** [comment]
> Can we suggest a filled-in array in the first place?

I'm sure we could, I haven't looked hard enough to know exactly how much work it would be. I saw this comment (maybe you wrote it?) and just assumed it would be more work than was worth it for now.

```
// TODO: ideally we'd gather deps here but it would require
// restructuring the rule code. This will cause a new lint
// error to appear immediately for useCallback. Note we're
// not adding [] because would that changes semantics.
```

**Comment by gaearon** [comment]
>Should we leave the autofix for functions on the assumption that it's not very common for folks to mutate functions?

Yeah I think that's reasonable to keep.

**Comment by captbaritone** [comment]
@gaearon  Actually, I'm trying to think how hard it would be to detect cases where the value _might_ have been mutated. Would it be sufficient to identify cases where the object has one of its properties read/written/called? If it's that simple, and if the ESLint scope API can give us that info easily (I'm still new to that API and it's poorly documented) then maybe it would be possible to allow the fix in the common case where the value is not mutated.

**Comment by gaearon** [comment]
One way to tell is to go through `references` and see what its usage is like. It might work. But keep in mind it might be tricky to determine. Like if an array is used with `arr.map()` for rendering, we wouldn't know with a simple heuristic if this mutates or not.

**Comment by captbaritone** [comment]
Updated:

* Only offer this autofix for functions -- which are unlikely to get mutated.
* Update the message to clarify that the entire construction of the value should get wrapped not just the declaration.

**Comment by captbaritone** [comment]
I'll punt on trying to detect potential ussages for now.

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js [hunk 3]
```diff
@@ -5377,7 +5377,7 @@ const tests = {
             `The 'handleNext' function makes the dependencies of ` +
             `useEffect Hook (at line 11) change on every render. ` +
             `Move it inside the useEffect callback. Alternatively, ` +
-            `wrap the 'handleNext' definition into its own useCallback() Hook.`,
+            `wrap the construction of 'handleNext' in its own useCallback() Hook.`,
```
Comment (gaearon) [inline packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:5380]: I don't think "construction" is a common term in this case. People don't think about declaring functions as "constructing" them.

**Comment by gaearon** [comment]
Gonna review this tomorrow morning.

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js (hunk 3 above)
Comment (captbaritone) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:5380]: Yeah, I wasn't 100% happy with this either. Can you think of a term which would encompass both declaration and any subsequent mutation needed to create an object/function? If the term needs to be different for objects/arrays than for functions that's fine.

File: packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js [hunk 4]
```diff
@@ -844,59 +844,77 @@ export default {
...
+            let suggest;
+            // Only handle the simple case of variable assignments.
+            // Wrapping function declarations can mess up hoisting.
+            if (
+              isUsedOutsideOfHook &&
+              construction.type === 'Variable' &&
+              // Objects may be mutated ater construction, which would make this
+              // fix unsafe. Functions _probably_ won't be mutated, so we'll
+              // allow this fix for them.
+              depType === 'function'
```
Comment (bradzacher) [inline packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js:885]: won't this only turn on the suggestion for functions?

File: packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js [hunk 5]
```diff
@@ -1381,50 +1399,97 @@ function collectRecommendations({
...
       if (
-        fnNode.type === 'FunctionName' &&
-        fnNode.node.type === 'FunctionDeclaration'
+        node.type === 'FunctionName' &&
+        node.node.type === 'FunctionDeclaration'
       ) {
-        return fnRef;
+        return [ref, 'function'];
       }
       return null;
```
Comment (bradzacher) [inline packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js:1485]: would be an exceedingly rare case, but could also add handling for class declarations:
```js
function Component() {
  class WhyTho {}
  useMemo(() => {}, [WhyTho]);
}
```

File: packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js [hunk 6]
```diff
@@ -1381,50 +1399,97 @@ function collectRecommendations({
...
+      return null;
+    case 'LogicalExpression':
+      if (
+        getConstructionExpresionType(node.left) != null ||
+        getConstructionExpresionType(node.right) != null
+      ) {
+        return 'logical expression';
+      }
+      return null;
+  }
```
Comment (bradzacher) [inline packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js:1431]: I believe that you could also add support for JSX here, because under the hood they are going to create a new object ref each render (right? I don't know the exact semantics of JSX transforms)

```js
const frag = <></>; // JSXFragment
const comp = <div />; // JSXElement
useMemo(() => {}, [frag, comp]);
```

Comment (bradzacher) [reply packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js:1431]: could also add handling for `AssignmentExpression`:
```js
let x = null;
let y = x = () => {};
useMemo(() => {}, [y]);
```

(i.e. ignore the `left` of the assignment)

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js (hunk 3 above)
Comment (gaearon) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:5380]: - wrap the [name] function body
- wrap the [name] object creation

?

File: packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js (hunk 6 above)
Comment (bradzacher) [reply packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js:1431]: could also add handling for `NewExpression`?
```js
const x = new Map();
useMemo(() => {}, [x]);
```

Comment (bradzacher) [reply packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js:1431]: could also add handling for `Literal` when `value instanceof RegExp`?
```js
const x = /a/;
useMemo(() => {}, [x]);
```

Comment (bradzacher) [reply packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js:1431]: could also handle TS type assertions (`TSAsExpression`):

```js
const x = {} as any;
useMemo(() => {}, [x]);
```

(i.e. ignore the node and check the `expression`.)

Comment (bradzacher) [reply packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js:1431]: could also handle flow type casts (`TypeCastExpression`):

```js
const x = ({}: any);
useMemo(() => {}, [x]);
```

(i.e. ignore the node and check the `expression`.)

File: packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js (hunk 4 above)
Comment (captbaritone) [reply packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js:885]: I hoped that the comment above would explain the reasoning.

File: packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js (hunk 6 above)
Comment (captbaritone) [reply packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js:1431]: This list is gold. Thanks! I had intentionally omitted `NewExpression` thinking that `String` `Number` and `Boolean` would be compared by value, but I forgot that they will be boxed so presumably they _will_ be referentially unique and therefore worth flagging.

Comment (bradzacher) [reply packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js:1431]: Yeah I didn't think about that, but testing in chrome shows they are referentially unique
image

Note that they're also not even loosely equal:

image

(note: if you swap one side for a literal, it coerces the object to the same type and then they become loosely equal)

File: packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js (hunk 4 above)
Comment (bradzacher) [reply packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js:885]: oops. I misread it. all good!

File: packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js [hunk 7]
```diff
@@ -844,59 +844,77 @@ export default {
...
+              depType === 'function'
+            ) {
+              suggest = [
+                {
+                  desc: `Wrap the construction of '${construction.name.name}' in its own ${wrapperHook}() Hook.`,
+                  fix(fixer) {
+                    const [before, after] =
+                      wrapperHook === 'useMemo'
+                        ? [`useMemo(() => { return `, '; })']
+                        : ['useCallback(', ')'];
```
[the remark this answers is not in the document]
Comment (captbaritone) [reply packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js:894]: Since this is only used within the actual fix, it seemed reasonable to have it here.

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js [hunk 8]
```diff
@@ -56,7 +56,7 @@ const tests = {
     {
       code: normalizeIndent`
         function MyComponent() {
-          const local = {};
+          const local = useMemo(() => ({}), []);
```
Comment (gaearon) [inline packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:59]: Can we just change these tests to be something like `const local = foo()`? To keep them simple and focused on one thing.

Applies to this one and all other tests below that don't specifically check the feature you're adding.

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js (hunk 3 above)
Comment (gaearon) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:5363]: This still needs to be changed.

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js (hunk 2 above)
Comment (gaearon) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:5775]: This still needs to be changed?

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js [hunk 9]
```diff
@@ -1376,6 +1376,64 @@ const tests = {
...
+      `,
+    },
+    {
+      code: normalizeIndent`
+        function useFoo() {
+          const foo = "fine";
+          if(true) {
+            const foo = {};
+          }
+          return useMemo(() => foo, [foo]);
```
Comment (gaearon) [inline packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:1417]: Add a comment for why this doesn't warn. (A limitation.) It's not obvious.

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js [hunk 10]
```diff
@@ -1376,6 +1376,64 @@ const tests = {
...
+          let [foo] = [1];
+          return useMemo(() => foo, [foo]);
+        }
+      `,
+    },
+    {
+      code: normalizeIndent`
+        function useFoo() {
+          const foo = "fine";
+          if(true) {
```
Comment (gaearon) [inline packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:1414]: Nit: space plz

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js (hunk 2 above)
Comment (captbaritone) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:5775]: Done in https://github.com/facebook/react/pull/19590/commits/e0ca072197821d999012773faf329d181851a2f4

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js [hunk 11]
```diff
@@ -6967,6 +6975,472 @@ const tests = {
...
+            "Move it inside the useMemo callback. Alternatively, wrap the construction of 'foo' in its own " +
+            'useMemo() Hook.',
+          suggestions: undefined,
+        },
+      ],
+    },
+    {
+      code: normalizeIndent`
+        function Component() {
+          const foo = true ? {} : "fine";
```
Comment (gaearon) [inline packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:7107]: Please add a "valid" test for `const foo = true ? "fine" : "fine"`. It's not obvious that this passes.

And if it doesn't pass, that's a problem. In that case it would be better not to consider conditionals at all than to fire false positives.

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js [hunk 12]
```diff
@@ -5735,63 +5784,63 @@ const tests = {
...
-                      console.log('hello');
-                    };
-                  }
-                  useEffect(() => {
-                    return Store.subscribe(handleNext);
-                  }, [handleNext]);
-                }
-              `,
+                "Wrap the construction of 'handleNext' in its own useCallback() Hook.",
+              output: normalizeIndent`
```
Comment (gaearon) [inline packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js]: Trailing spaces?

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js [hunk 13]
```diff
@@ -5735,63 +5784,63 @@ const tests = {
...
-            handleNext = () => {
-              console.log('hello');
-            };
-          }
-          useEffect(() => {
-            return Store.subscribe(handleNext);
-          }, [handleNext]);
-        }
+      code: normalizeIndent`	
+        function MyComponent(props) {
```
Comment (gaearon) [inline packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js]: Trailing spaces?

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js [hunk 14]
```diff
@@ -5805,6 +5796,56 @@ const tests = {
...
+          };	
+          if (props.foo) {	
+            handleNext = () => {	
+              console.log('hello');	
+            };	
+          }	
+          useEffect(() => {	
+            return Store.subscribe(handleNext);	
+          }, [handleNext]);	
+        }
```
Comment (bradzacher) [inline packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:5813]: nit - there are tabs at the end of all of these lines

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js [hunk 15]
```diff
@@ -7345,6 +7860,24 @@ const testsTypescript = {
         },
       ],
     },
+    {
+      code: normalizeIndent`
+        function Foo() {
+          const foo = {} as any;;
```
Comment (gaearon) [inline packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:7866]: Nit: double ;

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js (hunk 14 above)
Comment (captbaritone) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js]: https://github.com/facebook/react/pull/19590/commits/4dc35db7b77b345babba9146eb8ba40c2e2922f2

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js (hunk 3 above)
Comment (gaearon) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:5363]: I still think we should keep "definition" for functions. And maybe "assignment" for everything else.

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js [hunk 16]
```diff
@@ -5767,31 +5816,31 @@ const tests = {
...
-                    handleNext = () => {
-                      console.log('hello');
-                    };
-                  }
-                  useEffect(() => {
-                    return Store.subscribe(handleNext);
-                  }, [handleNext]);
-                }
-              `,
+          function MyComponent(props) {
```
Comment (gaearon) [inline packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:5830]: Indentation?

File: packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js [hunk 17]
```diff
@@ -1381,50 +1399,116 @@ function collectRecommendations({
   };
 }
 
-// Finds functions declared as dependencies
+// If the node will result in constructing a referentially unique value, return
+// its human readable type name, else return null.
+function getConstructionExpresionType(node) {
```
Comment (gaearon) [inline packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js:1404]: Nit: `Expression`

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js (hunk 9 above)
Comment (bradzacher) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:1417]: Note that this isn't a limitation - it's just not an error. The inner `const foo = {}` is a separate variable in a separate child scope, so its `{}` value is unrelated to the `const foo = "fine"` in the component's scope.

Though a comment explaining that would be worthwhile!

Comment (gaearon) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:1417]: Ahh.

**Comment by gaearon** [comment]
You'll want to add some tests for cases where the same variable is used by multiple `useMemo`'s.

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js (hunk 3 above)
Comment (captbaritone) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:5363]: I worry that suggesting that they "wrap the assignment" implies that its the right-hand-side of the assignment expression that needs to be wrapped, and does not convey that any subsequent mutations of that object _also_ need to get wrapped in the same `useMemo`.

Maybe it's the best we can do, so I'll go with that for now, but I wish there were a way to clearly communicate that given:

```
const foo = []; // <== Assignment
foo.push(bar);
```

We want them to wrap _both_ the assignment _and_ the mutation (like this):

```
// GOOD
const foo = useMemo(() => {
   const f = [];
   f.push(bar);
}, [bar]);
```

And not _just the assignment_ (like this): 

```
// BAD
const foo = useMemo(() => [], []);
foo.push(bar);
```

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js (hunk 11 above)
Comment (captbaritone) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:7107]: `const foo = true ? "fine" : "fine"` passes. I'll add a test. Good idea.

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js (hunk 3 above)
Comment (gaearon) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:5363]: "creation" then?

Comment (gaearon) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:5363]: "initialization" is also fine by me.

File: packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js (hunk 6 above)
Comment (gaearon) [reply packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js]: This is such an edge case btw. I don't think handling boxed primitives is super important.

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js (hunk 3 above)
Comment (captbaritone) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js]: Ah, I like initialization. I'll go with that.

Comment (captbaritone) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js]: Done in https://github.com/facebook/react/pull/19590/commits/2b7ae4b79faed0d182b0aaa468b79b63171dc437

File: packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js (hunk 6 above)
Comment (captbaritone) [reply packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js]: @gaearon I would agree that it's very unlikely to happen, we get it for free when handling other `NewExpression`s like `new Map([])`.

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js (hunk 8 above)
Comment (captbaritone) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js]: https://github.com/facebook/react/pull/19590/commits/eaa2aff9424e7ded164799fa2aef27ccfa4ed3d0

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js (hunk 9 above)
Comment (captbaritone) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:1418]: Added a test in https://github.com/facebook/react/pull/19590/commits/eaa2aff9424e7ded164799fa2aef27ccfa4ed3d0

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js (hunk 11 above)
Comment (captbaritone) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:7116]: Added here: https://github.com/facebook/react/pull/19590/commits/eaa2aff9424e7ded164799fa2aef27ccfa4ed3d0

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js (hunk 15 above)
Comment (captbaritone) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js]: Fixed in https://github.com/facebook/react/pull/19590/commits/eaa2aff9424e7ded164799fa2aef27ccfa4ed3d0

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js (hunk 16 above)
Comment (captbaritone) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js]: Fixed in https://github.com/facebook/react/pull/19590/commits/eaa2aff9424e7ded164799fa2aef27ccfa4ed3d0

File: packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js (hunk 17 above)
Comment (captbaritone) [reply packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js]: Fixed in https://github.com/facebook/react/pull/19590/commits/eaa2aff9424e7ded164799fa2aef27ccfa4ed3d0

**Comment by captbaritone** [comment]
@gaearon I've added a test for a constant construction used by multiple dependency arrays here: https://github.com/facebook/react/pull/19590/commits/31f75cd4a0579c0be93b71ffc48ce0b308dad334

Note that it creates two different errors at the same location. One referencing the first dependency array, and one referencing the other.

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js (hunk 10 above)
Comment (captbaritone) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js]: Fixed in https://github.com/facebook/react/pull/19590/commits/eaa2aff9424e7ded164799fa2aef27ccfa4ed3d0

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js (hunk 12 above)
Comment (captbaritone) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js]: https://github.com/facebook/react/pull/19590/commits/4dc35db7b77b345babba9146eb8ba40c2e2922f2

**Comment by gaearon** [comment]
For patterns where an object doesn't use any values from the pure scopes, would it be possible to change the message to suggest to move it *outside* the component scope? This is a lot less noisy and seems fairly common in practice.

**Comment by captbaritone** [comment]
> For patterns where an object doesn't use any values from the pure scopes, would it be possible to change the message to suggest to move it outside the component scope? This is a lot less noisy and seems fairly common in practice.

Good suggestion, but I think it would require significantly more analysis. Currently we don't do any work to try to understand the values used to construct the object. If we did, not only could we adjust the message as you suggest, but we could also construct the dependency array for the autofix of adding a wrapping `useMemo`. Given the amount of work involved, I think it makes sense to leave this for another task.

**Comment by nstepien** [comment]
Will it also handle cases like default props? e.g.:
```js
function Foo({ arr = [] }) {
  useMemo(() => {
    console.log(arr);
  }, [arr]);
}
```

**Comment by captbaritone** [comment]
@nstepien Good question. It doesn't currently. It looks like it wouldn't be too hard to add though. Here's a commit that adds it: https://github.com/facebook/react/commit/29d560bacb3045dea01de9c2e490e1f730e6c204

I'm happy to open that as a PR, but I suspect we would want to come up with a somewhat different error message in that case, and I'm not sure I have the time right now to work on that (maybe you do?).

Thoughts @gaearon? Do you agree that we would want to different error message for the default param case?

**Comment by gaearon** [comment]
That would need a different message, yes.

**Comment by nstepien** [comment]
@captbaritone Checking for default props is still missing, do you think you could work on it? I can log an issue otherwise.
It should also check for default params in custom hooks, i.e.:
```js
function useArr(arr = []) {
  useEffect(() => {
    console.log(arr);
  }, [arr]);
}
```
